### ISSUE TITLE
Iterable support, Map/Set/Immutable support, lazy values/sequences

### ISSUE THREAD

**Issue by jdeal** [jdeal/qim#5]
This is a continuation of https://github.com/jdeal/qim/issues/4, but that issue is kind of murky, so starting things fresh here.

I've been hacking for a while now on adding these things:

1. Support for lazy values/sequences. For example, `[$slice(0, 100), $set([])]` could avoid evaluating the slice since it's never actually used. (There's some hackety-hack lookahead right now to do that, but I'd like to remove that.) Similarly, `[$slice(0, 10), $each, 'isTopTen', $set(true)]` could avoid creating an intermediate array and simply pass along a sequence that could be lazy iterated by `$each`.
2. Abstracting away mutations. Right now, there's a lot of conditional slicing/splicing/assigning/etc, depending on whether something is an array or an object. Similar to Immutable.js's `asMutable`, I thought I could use the same wrapper (internally) for lazy sequences to be able to create mutable clones and make it a little more straightforward to creating navigators that do updates.
3. Support iterables like `Map` and `Set` and possibly even Immutable.js's collections. Again, the same wrapper would be handy for abstracting away different data types.

There are some problems with all this though:

1. Creating a wrapper for lazy evaluation is not performance-neutral. Although I'm not trying to deeply marshall objects like with Immutable.js, there's still a similar tradeoff. The cost of wrapping eats away at some of the benefit of wrapping.
2. A full lazy evaluating/iterating wrapper is actually really complex. So it adds a lot of complexity and extra size to the implementation. Look at [Lazy.js](https://github.com/dtao/lazy.js/blob/master/lazy.js) for example to get an idea where the slippery slope could lead.
3. The meaning of some of qim's core navigators gets kind of murky. For example, when should a string key be referencing an object property and when should it be referencing a key in a Map? When should we iterate over an object's iterable values vs its properties? Also, how do we know how to clone and mutate iterables?

I'll elaborate on that last point with an example, since it's why I decided to open this issue:

```js
const map = new Map([['a', 'foo'], ['b', 'bar'])
find(['a'], map)
```

It seems like that should get `'foo'` from our map. And this:

```js
update(['c', $set('baz')], map)
map.get('c')
```

should set the key `'c'` on our map to `'baz'`. There are already some problems though:

1. How do we _generically_ know how to get and set these properties? Do we look for `get` and `set` and `add` methods? What if those just happen to be setting on a random object? What if they have a different meaning on the random object?
2. How do we _generically_ know how to clone the object? It's probably pretty safe to call the constructor, but what if the object was constructed with certain parameters and we don't construct it that way?

Originally, I thought I'd look for `Symbol.iterator` on the object (if it's not an array) and assume that it was meant to be treated as an iterable and not as an object. So for example:

```js
let object = {props: {}};
// add an x property to object, which is fine
object = update(['x', $set(1)], object);
// add an iterator property
object[Symbol.iterator] = function () {
  // code for creating an iterator
};
// should this throw an error because there's no `set` method?
// or should it still treat this like a plain object?
object = update(['x', $set(1)], object);
// add a set method
object.set = function (k, v) => {
  object.props[k] = v;
};
// now this starts behaving differently?
object = update(['x', $set(1)], object);
```

There's a big question mark as to what `object` even is at the end there. If qim cloned the object, the clone would be holding onto the same `props` object, and any calls to `set` would keep mutating that `props` object. 😱 

That illustrates some of the somewhat arbitrary behavior you can get. This is potentially edgy, but I think it shows that trying to add magic do-the-right-thing-for-all-possible-iterables behavior gets pretty tricky.

I could instead just support Map and Set and forget all about generic support. But of course in JS, it's notoriously difficult (and possibly expensive) to definitively check the type of an object. (`Array.isArray` exists specifically because it's tricky to determine if an object is an array.)

I could also add some a special navigator like `$key` that uses `.get` if it exists. Because Map supports objects and not just strings, that's necessary anyway. For the case where someone is using only Maps though, that would get pretty unwieldy. That would effectively negate the desirability of using qim for Immutable.

To summarize this book, the options are:

1. Do nothing. Drop this idea to add any kind of abstraction layer around values/sequences.
2. Add lazy value/sequence support around plain objects/arrays only. Suffer some additional complexity and performance cost, but no weird 💩 to deal with.
3. Add generic iterable support, based on duck typing. Suffer more additional complexity and performance cost. Don't worry about weird 💩 happening when duck typing goes wrong.
4. Only support `Map` and `Set`. If this were simple, it might reduce the complexity and weirdness somewhat, but the simple case of checking the `prototype` breaks `extends` and has other potential problems.

**Comment by jdeal** [comment]
Closing this since #10 adds an abstraction layer around sequences in a pretty performant way. No attempt has been made to add in iterable support there, so it's still an unknown cost. But if there's a path forward, that PR is it.
